### ISSUE TITLE
Deadlock due to race condition in download thread when using Bigquery Storage API

### ISSUE THREAD

**Issue by kien-truong** [googleapis/python-bigquery#2032]
There is a race condition in the download thread and the main thread when using Bigquery Storage API to fetch data.

https://github.com/googleapis/python-bigquery/blob/a76af359525ef3c49c958663f81fd24c9d35e1e7/google/cloud/bigquery/_pandas_helpers.py#L797-L801

https://github.com/googleapis/python-bigquery/blob/a76af359525ef3c49c958663f81fd24c9d35e1e7/google/cloud/bigquery/_pandas_helpers.py#L934-L942

When the download thread is blocked on `worker_queue.put(item)`, if the main thread exit, causing the pool to shut down, the download thread will be stuck. This behavior prevents the program from exiting.

**Comment by kien-truong** [comment]
@chalmerlowe can you take a look at this issue and my corresponding PR #2034? thanks

**Comment by tswast** [comment]
FWIW: We think this was affecting the BigQuery DataFrames benchmark tests. Basically, if one starts iterating through the dataframe pages but then stops iterating through, the workers would stay producing messages and never go totally out of scope / shutdown.
